File: packages/owl_ads_gromore/ios/owl_ads_gromore/Sources/owl_ads_gromore/PendingReply.swift
import Foundation

/// One-shot Pigeon reply guarded by the native request generation.
final class PendingReply<Value> {
  private let generation: Int64
  private var completion: ((Result<Value, Error>) -> Void)?

  init(generation: Int64, completion: @escaping (Result<Value, Error>) -> Void) {
    self.generation = generation
    self.completion = completion
  }

  @discardableResult
  func complete(generation callbackGeneration: Int64, result: Result<Value, Error>) -> Bool {
    guard callbackGeneration == generation, let completion else { return false }
    self.completion = nil
    completion(result)
    return true
  }
}

File: packages/owl_ads_gromore/android/src/main/kotlin/com/owlllwo/plugins/gromore/PendingReply.kt
package com.owlllwo.plugins.gromore

/** One-shot Pigeon reply guarded by the native request generation. */
internal class PendingReply<T>(
    private val generation: Long,
    callback: (Result<T>) -> Unit,
) {
    private var callback: ((Result<T>) -> Unit)? = callback

    fun complete(callbackGeneration: Long, result: Result<T>): Boolean {
        if (callbackGeneration != generation) return false
        val pending = callback ?: return false
        callback = null
        pending(result)
        return true
    }
}
